196-UIKitの学習-11.Closure

keigoakiba opened the pull request:
動作確認はシミュレーターにてお願いします
https://app.shortcut.com/sapporomobile/story/196/uikit-11-closure

Commit:
APIコールを行なっているfetchWeather()に関数引数であるcompletionを追加
ViewControllerクラス内のupedateForecast()にてcompletionクロージャの処理内容を記述
Reloadボタン押下時とフォアグラウンド帰還時のメソッドの処理を一つの関数forecastHandler()にまとめ、それを呼び出すよう変更

keigoakiba commented:
※これまでの本成果物がdelegateパターンになっていなかったため、課題を独自変更し、
「APIの結果を、コールバック形式で受け取る」ようにできていればOKとしました

※また、Reloadボタンが押下された時とアプリがフォアグランドに戻ってきた時の処理内容が同じであるため、それらの処理を一つの関数 forecastHandler()としてまとめ、ViewControllerクラス内に新規作成しました


-前回からの変更点-
ViewController.swift
73行目 completion引数の追加
103行目 completion()の呼び出しを追加
42行目 completion引数の追加
154行目 completionクロージャの処理内容をトレイリング方式で記述し追加
183行目 forecastHandler()の追加
197,202行目 forecastHandler()を呼び出す方法に変更

taketomato commented on Training/Training/ViewController.swift:
```diff
@@ -178,30 +180,26 @@ class ViewController: UIViewController {
     
-    //バックグラウンドからフォアグラウンドに戻った際に実行される処理
-    @objc func handleWillEnterForeground(notification: Notification) {
+    func forecastHandler(_ forecast: ForecastProtocol?) {
```
共通化するにあたって、 `forecast` を引数にした理由はなんでしょうか？
このクラスのプロパティ(=インスタンス変数)として持っていてそれを使うのであれば、引数にする理由がないように見えました。

keigoakiba replied:
updatateForecast()やdisplayForecast()でもそうしたように、UTにてdefaultForecast以外の具象クラスを引数に与えてテストすることができると考えたためでした。
ですが、forecastHandlerを今後UTで引数アリで使う見込みがなかったので、引数なしに修正します。

taketomato commented on Training/Training/ViewController.swift:
```diff
@@ -150,9 +151,10 @@ class ViewController: UIViewController {
-        receiveInfo = forecast.receiveInfo
-        weatherIcon = forecast.getWeatherIcon()
+        forecast.fetchWeather {
+            receiveInfo = forecast.receiveInfo
```
Closure を使う恩恵を少しでも受けるためにも `ReceiveInfo` を Clocuse の戻り値にしてみましょうか。
戻り値にすると `YumemiForecast` のプロパティの `receiveInfo` が削除できます。

keigoakiba replied:
修正しました。

taketomato reviewed (changes requested):
UT が通らなくなっています。修正お願いします。

Commit:
forecastHandler()の引数を削除
YumemiForecast のプロパティの receiveInfo を削除し、fetchWeather()の変数として再定義
forecastModel.swiftの各クラスのfetchWeather()にcompletionを追加

Commit:
不要なコメント部分の削除

taketomato reviewed (approved):
修正OK でっす！！ 🎉
